File: binary_search_tree/binary_search_tree.py
class BST:
  def __init__(self, value):
    self.value = value
    self.left = None
    self.right = None

  # Average: O(log(n)) time | O(1) space
  # Worst: O(n) time | O(1) space
  def insert(self, value):
    current_node = self
    while True:
      if value < current_node.value:
        if current_node.left is None:
          current_node.left = BST(value)
          break
        else:
          current_node = current_node.left
      else:
        if current_node.right is None:
          current_node.right = BST(value)
          break
        else:
          current_node = current_node.right
    return self


  # Average: O(log(n)) time | O(1) space
  # Worst: O(n) time | O(1) space
  def contains(self, value):
    current_node = self
    while current_node is not None:
      if value < current_node.value:
        current_node = current_node.left
      elif value > current_node.value:
        current_node = current_node.right
      else:
        return True
    return False

  # Average: O(log(n)) time | O(1) space
  # Worst: O(n) time | O(1) space
  def remove(self, value, parent_node = None):
    current_node = self
    while current_node is not None:
      if value < current_node.value:
        parent_node = current_node
        current_node = current_node.left
      elif value > current_node.value:
        parent_node = current_node
        current_node = current_node.right
      else:
        if current_node.left is not None and current_node.right is not None:
          current_node.value = current_node.right.get_min_value()
          current_node.right.remove(current_node.value, current_node)
        elif parent_node is None:
          if current_node.left is not None:
            current_node.value = current_node.left.value
            current_node.right = current_node.left.right
            current_node.left = current_node.left.left
          elif current_node.right is not None:
            current_node.value = current_node.right.value
            current_node.left = current_node.right.left
            current_node.right = current_node.right.right
          else:
            pass
        elif parent_node.left == current_node:
          parent_node.left = current_node.left if current_node.left is not None else current_node.right
        elif parent_node.right == current_node:
          parent_node.right = current_node.left if current_node.left is not None else current_node.right
        break
    return self

  def get_min_value(self):
    current_node = self
    while current_node.left is not None:
      current_node = current_node.left
    return current_node.value

bst = BST(10)
bst.insert(5).insert(4).insert(6).insert(15).insert(12).insert(17)

print(bst.remove(10).contains(10))

File: binary_search_tree/binary_search_tree.ts
class BST {
  value: number;
  left: BST | null = null;
  right: BST | null = null;

  constructor(value: number) {
    this.value = value;
  }

  insert(value: number): this {
    let currentNode: BST = this;
    while (true) {
      if (value < currentNode.value) {
        if (!currentNode.left) {
          currentNode.left = new BST(value);
          break;
        } else {
          currentNode = currentNode.left;
        }
      } else {
        if (!currentNode.right) {
          currentNode.right = new BST(value);
          break;
        } else {
          currentNode = currentNode.right;
        }
      }
    }
    return this;
  }

  contains(value: number): boolean {
    let currentNode: BST | null = this;
    while (currentNode) {
      if (value < currentNode.value) {
        currentNode = currentNode.left;
      } else if (value > currentNode.value) {
        currentNode = currentNode.right;
      } else {
        return true;
      }
    }
    return false;
  }

  remove(value: number, parentNode: BST | null = null): this {
    let currentNode: BST | null = this;
    while (currentNode) {
      if (value < currentNode.value) {
        parentNode = currentNode;
        currentNode = currentNode.left;
      } else if (value > currentNode.value) {
        parentNode = currentNode;
        currentNode = currentNode.right;
      } else {
        if (currentNode.left && currentNode.right) {
          currentNode.value = currentNode.right.getMinValue();
          currentNode.right.remove(currentNode.value, currentNode);
        } else if (!parentNode) {
          if (currentNode.left) {
            currentNode.value = currentNode.left.value;
            currentNode.right = currentNode.left.right;
            currentNode.left = currentNode.left.left;
          } else if (currentNode.right) {
            currentNode.value = currentNode.right.value;
            currentNode.left = currentNode.right.left;
            currentNode.right = currentNode.right.right;
          } else {
            // Single node tree, do nothing
          }
        } else if (parentNode.left === currentNode) {
          parentNode.left = currentNode.left
            ? currentNode.left
            : currentNode.right;
        } else if (parentNode.right === currentNode) {
          parentNode.right = currentNode.left
            ? currentNode.left
            : currentNode.right;
        }
        break;
      }
    }
    return this;
  }

  getMinValue(): number {
    let currentNode: BST = this;
    while (currentNode.left) {
      currentNode = currentNode.left;
    }
    return currentNode.value;
  }
}

// Example usage
const bst = new BST(10);
bst.insert(5).insert(4).insert(6).insert(15).insert(12).insert(17);

bst.remove(10);
console.log(bst.contains(10)); // false
